apps/costcontrol/js/app.js

```diff
@@ -350,7 +350,9 @@ var CostControlApp = (function() {
         debug('App UI mode: ', mode);
 
         // Layout
-        if (mode !== currentMode) {
+        if (mode === currentMode) {
```

Comment:
I think, the else branch calls the callback in an asynchronous way. Could you try to call it synchronous (we always can turn it asynchronous from the caller) but at the end of the else branch. Even better, why not factor out this callback after the `if`?